gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveDataset.java

```diff
@@ -0,0 +1,261 @@
+  private final Table table;
+  private final Splitter splitter = Splitter.on(",").omitEmptyStrings();
+
+  private final List<Path> tableLocations;
+  // Only set if table has exactly one location
+  private final Optional<Path> tableRootPath;
+  private final Optional<Path> targetTableRoot;
+  private final boolean relocateDataFiles;
+
+  public HiveDataset(FileSystem fs, GenericObjectPool<IMetaStoreClient> clientPool, Table table, Properties properties) {
+    this.fs = fs;
+    this.clientPool = clientPool;
+    this.table = table;
+    this.properties = properties;
+
+    this.tableLocations = parseLocationIntoPaths(this.table.getSd().getLocation());
+
+    this.tableRootPath = this.tableLocations.size() == 1 ?
+        Optional.of(PathUtils.deepestNonGlobPath(this.tableLocations.get(0))) : Optional.<Path>absent();
+
+    this.relocateDataFiles =
+        Boolean.valueOf(properties.getProperty(RELOCATE_DATA_FILES_KEY, DEFAULT_RELOCATE_DATA_FILES));
+    this.targetTableRoot = properties.containsKey(COPY_TARGET_TABLE_ROOT) ?
+        Optional.of(resolvePath(properties.getProperty(COPY_TARGET_TABLE_ROOT))) : Optional.<Path>absent();
+  }
+
+  /**
+   * Takes a path with tokens {@link #databaseToken} or {@link #tableToken} and replaces these tokens with the actual
+   * database names and table name. For example, if db is myDatabase, table is myTable, then /data/$DB/$TABLE will be
+   * resolved to /data/myDatabase/myTable.
+   */
+  private Path resolvePath(String pattern) {
+    pattern = pattern.replaceAll(databaseToken, this.table.getDbName());
+    if (pattern.contains(tableToken)) {
+      pattern = pattern.replaceAll(tableToken, this.table.getTableName());
+      return new Path(pattern);
+    } else {
+      return new Path(pattern, this.table.getTableName());
+    }
+  }
+
+  /**
+   * Parse a location string from a {@link StorageDescriptor} into and actual list of globs.
+   */
+  private List<Path> parseLocationIntoPaths(String sdLocation) {
+    List<Path> locations = Lists.newArrayList();
+
+    if (sdLocation == null) {
+      return locations;
+    }
+
+    for (String location : this.splitter.split(sdLocation)) {
+      locations.add(new Path(location));
+    }
+    return locations;
+  }
+
+  /**
+   * Finds all files read by the table and generates CopyableFiles. The semantics are as follows:
+   * 1. Find all valid {@link StorageDescriptor}. If the table is partitioned, the {@link StorageDescriptor} of the base
+   *    table will be ignored, and we will instead process the {@link StorageDescriptor} of each partition.
+   * 2. For each {@link StorageDescriptor} find all files referred by it.
+   * 3. Generate a {@link CopyableFile} for each file referred by a {@link StorageDescriptor}.
+   * 4. If the table is partitioned, create a file set for each partition.
+   */
+  @Override public Collection<CopyableFile> getCopyableFiles(FileSystem targetFs, CopyConfiguration configuration)
```

Comment:
What if a table already exists in the target hive metastore? Why not comparing the src and target metadata on table/partition level?